File: 프로그래머스/lv2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque

def solution(maps):
    answer = 0
    dx=[-1,1,0,0]
    dy=[0,0,-1,1]
    q=deque()
    q.append([0,0])
    n=len(maps)
    m=len(maps[0])
    visited=[[0]*m for i in range(n)]
    
    while q:
        cx,cy=q.popleft()
        for i in range(4):
            nx,ny=cx+dx[i],cy+dy[i]
            
            if nx<0 or ny<0 or nx>=n or ny>=m:
                continue
            if visited[nx][ny]==0 and maps[nx][ny]==1:
                visited[nx][ny]=visited[cx][cy]+1
                q.append([nx,ny])
    
    if visited[n-1][m-1]==0:
        return -1
    else:
        return visited[n-1][m-1]+1

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;

class Solution {
    static int[][] visited;
    static int[] x = {-1, 1, 0, 0};
    static int[] y = {0, 0, -1, 1};
    
    public int solution(int[][] maps) {
        int answer = 0;
        visited = new int[maps.length][maps[0].length];
        
        bfs(maps);
        
        answer = visited[maps.length - 1][maps[0].length - 1];
        
        if (answer == 0) {
            return -1;
        } else {
            return answer;
        }
    }
    
    private void bfs(int[][] maps) {
        Queue<int[]> q = new LinkedList<>();
        visited[0][0] = 1;
        q.add(new int[]{0, 0});
        
        while (!q.isEmpty()) {
            int[] cur = q.poll();
            int cx = cur[0];
            int cy = cur[1];
            
            for (int i = 0; i < 4; i++) {
                int dx = cx + x[i];
                int dy = cy + y[i];
                
                if (dx < 0 || dy < 0 || dx >= maps.length || dy >= maps[0].length) {
                    continue;
                }
                
                if (maps[dx][dy] == 1 && visited[dx][dy] == 0) {
                    visited[dx][dy] = visited[cx][cy] + 1;
                    q.add(new int[]{dx, dy});
                }
            }
        }
    }
}